sinks/kafka/driver.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kafka
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/optiopay/kafka"
+	"github.com/optiopay/kafka/proto"
+	"k8s.io/heapster/extpoints"
+	sink_api "k8s.io/heapster/sinks/api"
+	kube_api "k8s.io/kubernetes/pkg/api"
+)
+
+const (
+	partition                = 0
+	brokerClientID           = "kafka-sink"
+	brokerDialTimeout        = 10 * time.Second
+	brokerDialRetryLimit     = 10
+	brokerDialRetryWait      = 500 * time.Millisecond
+	brokerAllowTopicCreation = true
+	brokerLeaderRetryLimit   = 10
+	brokerLeaderRetryWait    = 500 * time.Millisecond
+)
+
+type kafkaSink struct {
+	producer        kafka.Producer
+	timeSeriesTopic string
+	eventsTopic     string
+	sinkBrokerHosts []string
+}
+
+// START: ExternalSink interface implementations
+
+func (self *kafkaSink) Register(mds []sink_api.MetricDescriptor) error {
+	return nil
+}
+
+func (self *kafkaSink) Unregister(mds []sink_api.MetricDescriptor) error {
+	return nil
+}
+
+func (self *kafkaSink) StoreTimeseries(timeseries []sink_api.Timeseries) error {
+	if timeseries == nil || len(timeseries) <= 0 {
+		return nil
+	}
+	for _, t := range timeseries {
+		err := self.produceKafkaMessage(t, self.timeSeriesTopic)
+		if err != nil {
+			return fmt.Errorf("failed to produce Kafka messages: %s", err)
+		}
+	}
+	return nil
+}
+
+func (self *kafkaSink) StoreEvents(events []kube_api.Event) error {
+	if events == nil || len(events) <= 0 {
+		return nil
+	}
+	for _, event := range events {
+		err := self.produceKafkaMessage(event, self.eventsTopic)
+		if err != nil {
+			return fmt.Errorf("failed to produce Kafka messages: %s", err)
+		}
+	}
+	return nil
+}
+
+// produceKafkaMessage produces messages to kafka
+func (self *kafkaSink) produceKafkaMessage(v interface{}, topic string) error {
+	if v == nil {
+		return nil
+	}
+	jsonItems, err := json.Marshal(v)
+	if err != nil {
+		return fmt.Errorf("failed to transform the items to json : %s", err)
+	}
+	message := &proto.Message{Value: []byte(string(jsonItems))}
```

Comment:
SGTM. You can create a versioned structure in this sink.

On Mon, Oct 26, 2015 at 4:32 AM, yuqi huang <notifications@github.com>
wrote:

> In sinks/kafka/driver.go
> <https://github.com/kubernetes/heapster/pull/648#discussion_r42982284>:
>
> > +			return fmt.Errorf("failed to produce Kafka messages: %s", err)
> > +		}
> > +	}
> > +	return nil
> > +}
> > +
> > +// produceKafkaMessage produces messages to kafka
> > +func (self *kafkaSink) produceKafkaMessage(v interface{}, topic string) error {
> > +	if v == nil {
> > +		return nil
> > +	}
> > +	jsonItems, err := json.Marshal(v)
> > +	if err != nil {
> > +		return fmt.Errorf("failed to transform the items to json : %s", err)
> > +	}
> > +	message := &proto.Message{Value: []byte(string(jsonItems))}
>
> Thanks for your comments. @vishh <https://github.com/vishh> .
> I need all the ops information, includes metrics and events.
> Exposing an internal data structure TimeSeries over the wire is not well
> advised. I will pick out the effective information about metrics and events
> to give the kafka sink.Include: timestamp, value of metrics and events.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/heapster/pull/648/files#r42982284>.
>